algebras/algebra/src/main/scala/endpoints/algebra/Requests.scala

```diff
@@ -112,6 +112,14 @@ trait Requests extends Urls with Methods with SemigroupalSyntax {
 
   /**
     * Request with a `String` body.
+    *
+    *   - Server interpreters accept requests with content-type `text/plain` and
+    *     reject requests with an incorrect content-type.
+    *   - Server interpreters will use the character encoding set in the
+    *     content-type header to determine how the text is decoded.
+    *   - Client interpreters supply content-type `text/plain` with an explicit
+    *     character encoding
+    *
```

Comment:
This is a slightly weaker variant than https://github.com/julienrf/endpoints/pull/539#discussion_r409044736,  since otherwise I'm not sure how to implement for Play or Http4s. The key difference is that instead of specifying that server interpreters will only require content-type when there are multiple request entity types, we leave it up to particular server implementations.

@julienrf is that a reasonable position?

Reply:
Does that mean that requests with no content-type might be rejected by some interpreters?

Reply:
Correct.

Right now, those server interpreters are Http4s and Play. Here's why:

  * Http4s [`EntityDecoder.decode`](https://github.com/http4s/http4s/blob/b8a6d9d2a5ddbf1652a78db30c3a613cec649688/core/src/main/scala/org/http4s/EntityDecoder.scala#L25) only looks at the content-type if `strict = true`, but then it also rejects requests that don't specify content type

  * Play's logic is a bit more spread out, but whenever a `BodyParser` looks at a content-type, it fails the request if the content-type is absent (see [the `text` parser](https://github.com/playframework/playframework/blob/67972edaa52057108eb1f0d108ffa1c204df4296/core/play/src/main/scala/play/api/mvc/BodyParsers.scala#L525), for instance)

We probably could re-implement the desired content-type logic all in `endpoints` (so that no content-type is still allowed). However, I see the value _endpoints_ as providing some re-usable minimal interface over existing well-tested servers/clients. Diving too deep into entity parsing would defeat that.

Reply:
> We probably could re-implement the desired content-type logic all in `endpoints` (so that no content-type is still allowed). However, I see the value _endpoints_ as providing some re-usable minimal interface over existing well-tested servers/clients. Diving too deep into entity parsing would defeat that.

I understand, but I hope this wouldn’t make our server implementations too strict :-/

I don’t know how much this can be a problem in practice, though. I just remember having been a bit annoyed by having to add a content-type because my request was otherwise rejected by a server. But this is not a big deal… I believe most HTTP clients automatically set the request content-type?

Otherwise I can surely help getting a more “tolerant” entity parsing in the Play server interpreter. For http4s I’m not sure but I can try.

Reply:
Yeah, most HTTP clients should set content-type (ours do for sure now).

> Otherwise I can surely help getting a more “tolerant” entity parsing in the Play server interpreter. For http4s I’m not sure but I can try.

Play has [`tolerantText`](https://github.com/playframework/playframework/blob/67972edaa52057108eb1f0d108ffa1c204df4296/core/play/src/main/scala/play/api/mvc/BodyParsers.scala#L463). However, it takes the position that if you are going to check content-type at all, you should also assert that it is present (ex: `tolerantText` doesn't assert content type `text/plain`, but it doesn't guard against `application/json`).

Reply:
I think that for both http4s and Play it should be feasible (and not too complex) to implement our own `EntityDecoder` and `BodyParser` (respectively), to behave such that they reject content-types that don’t match but accept requests with no content-type set. I can give it a try if you want!

Reply:
Here is an example of implementation of `textRequest` for http4s and Play that rejects incompatible content-types but accepts requests with no Content-Type header: https://github.com/julienrf/endpoints/commit/0ac804848e8ee9795aef5621c23ebd916f279250

But I’m still not sure what we should do.

BTW, I think you discovered some bugs: what happens currently if one sends a request with an incompatible Content-Type? It gets rejected by the Play server interpreter, but not with a “good” response. The default behavior of Play is to return an HTML response.

Reply:
> BTW, I think you discovered some bugs: what happens currently if one sends a request with an incompatible Content-Type? It gets rejected by the Play server interpreter, but not with a “good” response. The default behavior of Play is to return an HTML response.

Not sure about what the expected body type of the response type is, but I think the status 415 is correct (https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/415). I don't think this should be a 400 (I'm looking at the test case output that changed in the linked commit).

> Here is an example of implementation of `textRequest` for http4s and Play that rejects incompatible content-types but accepts requests with no Content-Type header: 0ac8048
>
> But I’m still not sure what we should do.

Aside from the 415 status code, these implementations generally make sense. My personal preference is still against this direction though.

As I see it, _endpoints_ interpreters should try to produce outputs that are fairly idiomatic for whatever framework/library they are targeting. Re-implementing logic in _endpoints_  makes it harder for the interpreted outputs (routes in this case) to fit neatly into the rest of the framework.

For instance, one of the cool reasons to use Play is to dynamically adjust the sort of errors that you want all routes to return (see https://www.playframework.com/documentation/2.5.x/ScalaErrorHandling). With the implementation you propose, routes created by _endpoints_ with `textRequest` would no longer obey this.

Reply:
Yes, 415 is a correct status code, I agree. The motivation for not letting Play handle errors is that an important property when using _endpoints_ is to have client, server, and documentation consistent together. If we let the underlying HTTP library handle errors, then this one is out of our control, and clients might receive error responses they can’t decode (or, which are not documented). For instance, an error response with an unknown status code, or worse, with an unexpected content-type (such HTML instead of JSON).

This is what led me to the current design that bypasses Akka-HTTP rejection handlers, and Play’s HttpErrorHandler.

Reply:
I still don't agree with content-type check, but I understand the motivation. Thanks for explaining! 👍

Reply:
> I still don't agree with content-type check

By “content-type check”, you mean the fact that in the commit I pushed I only check that the request content-type is not different than expected?

Reply:
Yeah sort of, although I don't disapprove of the code itself 😄.

I just disagree with the idea of trying to re-build content-type checking for servers with logic inside of _endpoints_. My opinion is we should defer to the server's own logic whenever possible.

Reply:
> I just disagree with the idea of trying to re-build content-type checking for servers with logic inside of _endpoints_. My opinion is we should defer to the server's own logic whenever possible.

I understand this point and I’m flexible to remove this custom logic. However, I still strongly believe that we should handle errors ourselves (not let the underlying library do it) so that we can guarantee that clients, servers, and documentation are consistent with the actual implementation. I’m talking here about the error that happens if a request is sent with an incompatible content-type. In such a case, we might want to return our own 415 error response (which needs to be properly documented), or we can just reuse the 400 error response (I know it’s not RESTful, but I sincerely doubt this can cause any real problem).

Reply:
Yeah, this makes sense. I do think respecting HTTP conventions like 415 is important, especially when building servers that are going to be consumed by third parties (who may not be relying on _endpoints_).

It would indeed be better for endpoints to uniformly manage all possible errors. This is a bit tricky in practice though. Some errors, such as 404, 405, (maybe even 415) require a more global view of the set of routes defined. Other errors, such as 413, 414, 429, 431 are status codes that I generally expect any server may return (depending on their operating constraints), but I'm not sure it would be useful for _endpoints_ to track them all.

In summary: I agree with your goal, but I worry doing it right will require very deep integration with servers. Personally I am satisfied if _endpoints_ focuses on clearly specifying the correct way to call REST endpoints and providing only a best-effort description/handling of possible errors (maybe with better guarantees around status codes 400 and 500).

Reply:
> I do think respecting HTTP conventions like 415 is important, especially when building servers that are going to be consumed by third parties (who may not be relying on _endpoints_).

I’ve always thought that respecting such conventions was be useful for humans only, but I doubt programs do have a special treatment when they receive a 415 error to e.g. try with a different content type, do they?

Reply:
In the case of 415, I agree there isn't much a program can do, but a status code can still inform how to present the error message. I can imagine a smart client getting a 413 (Request entity too large) and automatically trying to apply some tricks to retry with a compressed query. 401 is definitely useful for machines too.

Reply:
If we return a 415 status this should be included as possible response type in the documentation (like we do with statuses 500 and 400).

Reply:
Let’s move forward with this PR, though: what do you think of taking the path of requiring the `Content-Type` to be mandatory? (so that we can then implement #529)

Reply:
Yeah, I can do that. Before I do though:

Are you sure we should? We can move forward with #529 even without mandatory `Content-Type` (ie, the PR as it is right now). I was envisioning that `Content-Type` would only become absolutely mandatory for `RequestEntities` produced by `.orElse`.

Reply:
Right! OK, let’s live with the few differences of behaviors between each interpreter then.